File: aws-globalaccelerator-endpointgroup/src/test/java/software/amazon/globalaccelerator/endpointgroup/UpdateHandlerTest.java
package software.amazon.globalaccelerator.endpointgroup;

import com.amazonaws.services.globalaccelerator.model.Accelerator;
import com.amazonaws.services.globalaccelerator.model.AcceleratorStatus;
import com.amazonaws.services.globalaccelerator.model.DescribeAcceleratorRequest;
import com.amazonaws.services.globalaccelerator.model.DescribeAcceleratorResult;
import com.amazonaws.services.globalaccelerator.model.DescribeEndpointGroupRequest;
import com.amazonaws.services.globalaccelerator.model.DescribeEndpointGroupResult;
import com.amazonaws.services.globalaccelerator.model.EndpointGroup;
import com.amazonaws.services.globalaccelerator.model.EndpointGroupNotFoundException;
import com.amazonaws.services.globalaccelerator.model.UpdateEndpointGroupRequest;
import com.amazonaws.services.globalaccelerator.model.UpdateEndpointGroupResult;
import junit.framework.Assert;
import lombok.val;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.function.Executable;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import software.amazon.cloudformation.proxy.AmazonWebServicesClientProxy;
import software.amazon.cloudformation.proxy.HandlerErrorCode;
import software.amazon.cloudformation.proxy.Logger;
import software.amazon.cloudformation.proxy.OperationStatus;
import software.amazon.cloudformation.proxy.ProgressEvent;
import software.amazon.cloudformation.proxy.ResourceHandlerRequest;
import java.util.ArrayList;
import static org.mockito.Mockito.mock;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.doReturn;
import static org.mockito.Mockito.doThrow;
import static org.mockito.Mockito.mock;

@ExtendWith(MockitoExtension.class)
public class UpdateHandlerTest {

    @Mock
    private AmazonWebServicesClientProxy proxy;

    @Mock
    private Logger logger;

    @BeforeEach
    public void setup() {
        proxy = mock(AmazonWebServicesClientProxy.class);
        logger = mock(Logger.class);
    }

    private ResourceModel createTestResourceModel() {
        val endpointConfigurations = new ArrayList<EndpointConfiguration>();
        endpointConfigurations.add(EndpointConfiguration.builder().endpointId("EPID1").build());
        endpointConfigurations.add(EndpointConfiguration.builder().endpointId("EPID2").build());

        return ResourceModel.builder()
                .endpointGroupArn("ENDPOINTGROUP_ARN")
                .endpointGroupRegion("us-west-2")
                .listenerArn("LISTENER_ARN")
                .healthCheckPort(10)
                .thresholdCount(100)
                .endpointConfigurations(endpointConfigurations)
                .healthCheckPath("/Health")
                .trafficDialPercentage(100)
                .healthCheckIntervalSeconds(10)
                .healthCheckProtocol("TCP")
                .listenerArn("arn:aws:globalaccelerator::474880776455:accelerator/abcd1234/listener/12341234")
                .build();
    }

    @Test
    public void handleRequest_NonExistingEndpointGroup() {
        final UpdateHandler handler = new UpdateHandler();
        ResourceModel desiredModel = createTestResourceModel();
        doThrow(EndpointGroupNotFoundException.class).when(proxy).injectCredentialsAndInvoke(any(DescribeEndpointGroupRequest.class), any());

        final ResourceHandlerRequest<ResourceModel> request = ResourceHandlerRequest.<ResourceModel>builder()
                .desiredResourceState(desiredModel)
                .build();

        val response = handler.handleRequest(proxy, request, null, logger);
        Assertions.assertNotNull(response);
        Assertions.assertEquals(response.getStatus(), OperationStatus.FAILED);
        Assertions.assertEquals(response.getErrorCode(), HandlerErrorCode.NotFound);
        Assertions.assertNull(response.getResourceModel());
    }

    @Test
    public void handleRequest_UpdateEndpointGroup_PendingReturnsInProgress() {
        final UpdateHandler handler = new UpdateHandler();
        ResourceModel desiredModel = createTestResourceModel();

        val describeEndpointGroupResult = new DescribeEndpointGroupResult()
                .withEndpointGroup(new EndpointGroup()
                        .withEndpointGroupArn("ENDPOINTGROUP_ARN"));

        UpdateEndpointGroupResult updateEndpointGroupResult = new UpdateEndpointGroupResult()
                .withEndpointGroup(new EndpointGroup().withEndpointGroupArn("ENDPOINTGROUP_ARN"));

        doReturn(describeEndpointGroupResult).when(proxy).injectCredentialsAndInvoke(any(DescribeEndpointGroupRequest.class), any());
        doReturn(updateEndpointGroupResult).when(proxy).injectCredentialsAndInvoke(any(UpdateEndpointGroupRequest.class), any());

        final ResourceHandlerRequest<ResourceModel> request = ResourceHandlerRequest.<ResourceModel>builder()
                .desiredResourceState(desiredModel)
                .build();

        final ProgressEvent<ResourceModel, CallbackContext> response
                = handler.handleRequest(proxy, request, null, logger);

        Assertions.assertNotNull(response);
        Assertions.assertEquals(response.getStatus(), OperationStatus.IN_PROGRESS);
        Assertions.assertNotNull(response.getResourceModel());
        Assertions.assertNotNull(response.getCallbackContext());
        Assertions.assertEquals(response.getCallbackDelaySeconds(), 0);
        Assertions.assertEquals(response.getResourceModel().getEndpointGroupArn(), "ENDPOINTGROUP_ARN");

        Assertions.assertTrue(response.getCallbackContext().isPendingStabilization());

    }

    @Test
    public void handleRequest_UpdateEndpointGroup_DeployedReturnsSuccess() {
        final UpdateHandler handler = new UpdateHandler();
        ResourceModel desiredModel = createTestResourceModel();

        val describeEndpointGroupResult = new DescribeEndpointGroupResult()
                .withEndpointGroup(new EndpointGroup()
                        .withEndpointGroupArn("ENDPOINTGROUP_ARN"));

        val describeAcceleratorResult = new DescribeAcceleratorResult()
                .withAccelerator(new Accelerator()
                        .withAcceleratorArn("ACCELERATOR_ARN")
                        .withStatus(AcceleratorStatus.DEPLOYED.toString()));

        doReturn(describeAcceleratorResult).when(proxy).injectCredentialsAndInvoke(any(DescribeAcceleratorRequest.class), any());
        doReturn(describeEndpointGroupResult).when(proxy).injectCredentialsAndInvoke(any(DescribeEndpointGroupRequest.class), any());

        final ResourceHandlerRequest<ResourceModel> request = ResourceHandlerRequest.<ResourceModel>builder()
                .desiredResourceState(desiredModel)
                .build();

        CallbackContext context = new CallbackContext();
        context.setStabilizationRetriesRemaining(100);
        context.setPendingStabilization(true);

        final ProgressEvent<ResourceModel, CallbackContext> response
                = handler.handleRequest(proxy, request, context, logger);

        Assertions.assertNotNull(response);
        Assertions.assertEquals(response.getStatus(), OperationStatus.SUCCESS);
        Assertions.assertNotNull(response.getResourceModel());
        Assertions.assertNull(response.getCallbackContext());
        Assertions.assertEquals(response.getCallbackDelaySeconds(), 0);
        Assertions.assertEquals(response.getResourceModel().getEndpointGroupArn(), "ENDPOINTGROUP_ARN");
    }

    @Test
    public void testStabilizationTimeout() {
        final UpdateHandler handler = new UpdateHandler();
        ResourceModel desiredModel = createTestResourceModel();

        final ResourceHandlerRequest<ResourceModel> request = ResourceHandlerRequest.<ResourceModel>builder()
                .desiredResourceState(desiredModel)
                .build();

        val describeEndpointGroupResult = new DescribeEndpointGroupResult()
                .withEndpointGroup(new EndpointGroup()
                        .withEndpointGroupArn("ENDPOINTGROUP_ARN"));
        doReturn(describeEndpointGroupResult).when(proxy).injectCredentialsAndInvoke(any(DescribeEndpointGroupRequest.class), any());

        CallbackContext context = new CallbackContext();
        context.setStabilizationRetriesRemaining(0);
        context.setPendingStabilization(true);

        val excThrown = Assertions.assertThrows(RuntimeException.class, () -> handler.handleRequest(proxy, request, context, logger));
        excThrown.getMessage().contains("Timed out waiting for endpoint group to be deployed.");
    }

}

File: aws-globalaccelerator-endpointgroup/src/test/kotlin/software/amazon/globalaccelerator/endpointgroup/UpdateHandlerTest.kt
package software.amazon.globalaccelerator.endpointgroup

import com.amazonaws.services.globalaccelerator.model.Accelerator
import com.amazonaws.services.globalaccelerator.model.AcceleratorStatus
import com.amazonaws.services.globalaccelerator.model.DescribeAcceleratorRequest
import com.amazonaws.services.globalaccelerator.model.DescribeAcceleratorResult
import com.amazonaws.services.globalaccelerator.model.DescribeEndpointGroupRequest
import com.amazonaws.services.globalaccelerator.model.DescribeEndpointGroupResult
import com.amazonaws.services.globalaccelerator.model.EndpointGroup
import com.amazonaws.services.globalaccelerator.model.EndpointGroupNotFoundException
import com.amazonaws.services.globalaccelerator.model.UpdateEndpointGroupRequest
import com.amazonaws.services.globalaccelerator.model.UpdateEndpointGroupResult
import io.mockk.MockKAnnotations
import io.mockk.every
import io.mockk.verify
import io.mockk.slot
import io.mockk.impl.annotations.MockK
import io.mockk.junit5.MockKExtension
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Assertions.assertNull
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import software.amazon.cloudformation.proxy.AmazonWebServicesClientProxy
import software.amazon.cloudformation.proxy.HandlerErrorCode
import software.amazon.cloudformation.proxy.Logger
import software.amazon.cloudformation.proxy.OperationStatus
import software.amazon.cloudformation.proxy.ResourceHandlerRequest
import java.util.function.Function
import kotlin.collections.ArrayList

@ExtendWith(MockKExtension::class)
class UpdateHandlerTest {

    @MockK
    lateinit var proxy: AmazonWebServicesClientProxy

    @MockK(relaxed = true)
    lateinit var logger: Logger

    @BeforeEach
    fun setup() = MockKAnnotations.init(this)

    private val endpointGroupRegion = "us-west-2"
    private val acceleratorArn = "arn:aws:globalaccelerator::444607872184:accelerator/88127aa5-01d8-484c-80a0-349daaefce1d"
    private val listenerArn = "arn:aws:globalaccelerator::444607872184:accelerator/88127aa5-01d8-484c-80a0-349daaefce1d/listener/ee7358c2"
    private val endpointGroupArn = "arn:aws:globalaccelerator::444607872184:accelerator/88127aa5-01d8-484c-80a0-349daaefce1d/listener/ee7358c2/endpoint-group/de69a4b45005"


    private fun createTestResourceModel(): ResourceModel {
        val endpointConfigurations = ArrayList<EndpointConfiguration>()
        endpointConfigurations.add(EndpointConfiguration.builder().endpointId("EPID1").weight(100).clientIPPreservationEnabled(true).build())
        endpointConfigurations.add(EndpointConfiguration.builder().endpointId("EPID2").weight(100).clientIPPreservationEnabled(true).build())

        return ResourceModel.builder()
                .endpointGroupArn(endpointGroupArn)
                .endpointGroupRegion(endpointGroupRegion)
                .listenerArn(listenerArn)
                .healthCheckPort(10)
                .thresholdCount(100)
                .endpointConfigurations(endpointConfigurations)
                .healthCheckPath("/Health")
                .trafficDialPercentage(100.0)
                .healthCheckIntervalSeconds(10)
                .healthCheckProtocol("TCP")
                .build()
    }

    @Test
    fun handleRequest_NonExistingEndpointGroup() {
        val desiredModel = createTestResourceModel()
        every { proxy.injectCredentialsAndInvoke(ofType(), ofType<ProxyDescribeEndpointGroup>()) } throws EndpointGroupNotFoundException("NOT FOUND")

        val request = ResourceHandlerRequest.builder<ResourceModel>()
                .desiredResourceState(desiredModel)
                .build()

        val response = UpdateHandler().handleRequest(proxy, request, null, logger)
        assertNotNull(response)
        assertEquals(OperationStatus.FAILED, response.status)
        assertEquals(HandlerErrorCode.NotFound, response.errorCode)
        assertEquals(desiredModel, response.resourceModel)
    }

    @Test
    fun handleRequest_UpdateEndpointGroup_PendingReturnsInProgress() {
        val previousModel = createTestResourceModel()
        val desiredModel = createTestResourceModel()

        val describeEndpointGroupResult = DescribeEndpointGroupResult()
                .withEndpointGroup(EndpointGroup()
                        .withEndpointGroupArn(endpointGroupArn))
        every { proxy.injectCredentialsAndInvoke(ofType(), ofType<ProxyDescribeEndpointGroup>()) } returns describeEndpointGroupResult

        val updateEndpointGroupResult = UpdateEndpointGroupResult()
                .withEndpointGroup(EndpointGroup().withEndpointGroupArn(endpointGroupArn))
        every { proxy.injectCredentialsAndInvoke(ofType(), ofType<ProxyUpdateEndpointGroup>()) } returns updateEndpointGroupResult

        val request = ResourceHandlerRequest.builder<ResourceModel>()
                .desiredResourceState(desiredModel)
                .previousResourceState(previousModel)
                .build()

        val response = UpdateHandler().handleRequest(proxy, request, null, logger)

        assertNotNull(response)
        assertEquals(OperationStatus.IN_PROGRESS, response.status)
        assertNotNull(response.resourceModel)
        assertNotNull(response.callbackContext)
        assertEquals(0, response.callbackDelaySeconds)
        assertEquals(endpointGroupArn, response.resourceModel.endpointGroupArn)
        assertTrue(response.callbackContext!!.pendingStabilization)
    }

    @Test
    fun handleRequest_UpdateEndpointGroup_DeployedReturnsSuccess() {
        val desiredModel = createTestResourceModel()

        val describeAcceleratorResult = DescribeAcceleratorResult()
                .withAccelerator(Accelerator()
                        .withAcceleratorArn(acceleratorArn)
                        .withStatus(AcceleratorStatus.DEPLOYED.toString()))
        every { proxy.injectCredentialsAndInvoke(ofType(), ofType<ProxyDescribeAccelerator>()) } returns describeAcceleratorResult

        val request = ResourceHandlerRequest.builder<ResourceModel>()
                .desiredResourceState(desiredModel)
                .build()

        val context = CallbackContext(100, true)

        val response = UpdateHandler().handleRequest(proxy, request, context, logger)

        assertNotNull(response)
        assertEquals(OperationStatus.SUCCESS, response.status)
        assertNotNull(response.resourceModel)
        assertNull(response.callbackContext)
        assertEquals(0, response.callbackDelaySeconds)
        assertEquals(endpointGroupArn, response.resourceModel.endpointGroupArn)
    }

    @Test
    fun testStabilizationTimeout() {
        val desiredModel = createTestResourceModel()

        val request = ResourceHandlerRequest.builder<ResourceModel>()
                .desiredResourceState(desiredModel)
                .build()

        val context = CallbackContext(0, true)

        val handler = UpdateHandler()
        val exception = assertThrows(RuntimeException::class.java) {
            handler.handleRequest(proxy, request, context, logger)
        }
        assertEquals("Timed out waiting for endpoint group to be deployed.", exception.message)
    }

    @Test
    fun handleRequest_UpdateEndpointGroupWithPortOverrides() {
        val previousModel = createTestResourceModel()
        val desiredModel = createTestResourceModel()
        val portOverrides = listOf(PortOverride(80, 8080))
        desiredModel.portOverrides = portOverrides

        val describeEndpointGroupResult = DescribeEndpointGroupResult()
                .withEndpointGroup(EndpointGroup()
                        .withEndpointGroupArn(endpointGroupArn))
        every { proxy.injectCredentialsAndInvoke(ofType(), ofType<ProxyDescribeEndpointGroup>()) } returns describeEndpointGroupResult

        val updateEndpointGroupResult = UpdateEndpointGroupResult()
                .withEndpointGroup(EndpointGroup().withEndpointGroupArn(endpointGroupArn))
        every { proxy.injectCredentialsAndInvoke(ofType(), ofType<ProxyUpdateEndpointGroup>()) } returns updateEndpointGroupResult

        val request = ResourceHandlerRequest.builder<ResourceModel>()
                .desiredResourceState(desiredModel)
                .previousResourceState(previousModel)
                .build()

        val response = UpdateHandler().handleRequest(proxy, request, null, logger)

        assertNotNull(response)
        assertEquals(OperationStatus.IN_PROGRESS, response.status)
        assertNotNull(response.resourceModel)
        assertNotNull(response.callbackContext)
        assertEquals(0, response.callbackDelaySeconds)
        assertEquals(endpointGroupArn, response.resourceModel.endpointGroupArn)
        assertTrue(response.callbackContext!!.pendingStabilization)
        assertEquals(1, response.resourceModel.portOverrides.size)
        assertEquals(portOverrides[0].listenerPort, response.resourceModel.portOverrides[0].listenerPort)
        assertEquals(portOverrides[0].endpointPort, response.resourceModel.portOverrides[0].endpointPort)
    }

    @Test
    fun handleRequest_UpdateEndpointGroupWithNullWithPreviousPortOverrides() {
        val portOverrides = listOf(PortOverride(80, 8080))
        val previousModel = createTestResourceModel()
        previousModel.portOverrides = portOverrides

        val desiredModel = createTestResourceModel()
        val describeEndpointGroupResult = DescribeEndpointGroupResult()
                .withEndpointGroup(EndpointGroup()
                        .withEndpointGroupArn(endpointGroupArn))
        every { proxy.injectCredentialsAndInvoke(ofType(), ofType<ProxyDescribeEndpointGroup>()) } returns describeEndpointGroupResult

	val updateEndpointGroupRequestSlot = slot<UpdateEndpointGroupRequest>()
        val updateEndpointGroupResult = UpdateEndpointGroupResult()
                .withEndpointGroup(EndpointGroup().withEndpointGroupArn(endpointGroupArn))
        every { proxy.injectCredentialsAndInvoke(capture(updateEndpointGroupRequestSlot), ofType<ProxyUpdateEndpointGroup>()) } returns updateEndpointGroupResult

        val request = ResourceHandlerRequest.builder<ResourceModel>()
                .desiredResourceState(desiredModel)
                .previousResourceState(previousModel)
                .build()

        val response = UpdateHandler().handleRequest(proxy, request, null, logger)

        assertNotNull(response)
        assertEquals(response.status, OperationStatus.IN_PROGRESS)
        assertNotNull(response.resourceModel)
        assertNotNull(response.callbackContext)
        assertEquals(response.callbackDelaySeconds, 0)
        assertEquals(response.resourceModel.endpointGroupArn, endpointGroupArn)
        assertTrue(response.callbackContext!!.pendingStabilization)
        assertNull(response.resourceModel.portOverrides)
        verify(exactly = 1) { proxy.injectCredentialsAndInvoke(ofType(), ofType<ProxyUpdateEndpointGroup>()) }
        assertEquals(ArrayList<String>(), updateEndpointGroupRequestSlot.captured.portOverrides) // Empty array to clean port-overrides.
    }

    @Test
    fun handleRequest_UpdateEndpointGroupWithNullWithPreviousNullPortOverrides() {
        val previousModel = createTestResourceModel()
        val desiredModel = createTestResourceModel()

        val describeEndpointGroupResult = DescribeEndpointGroupResult()
                .withEndpointGroup(EndpointGroup()
                        .withEndpointGroupArn(endpointGroupArn))
        every { proxy.injectCredentialsAndInvoke(ofType(), ofType<ProxyDescribeEndpointGroup>()) } returns describeEndpointGroupResult

        val updateEndpointGroupRequestSlot = slot<UpdateEndpointGroupRequest>()
        val updateEndpointGroupResult = UpdateEndpointGroupResult()
                .withEndpointGroup(EndpointGroup().withEndpointGroupArn(endpointGroupArn))
        every { proxy.injectCredentialsAndInvoke(capture(updateEndpointGroupRequestSlot), ofType<ProxyUpdateEndpointGroup>()) } returns updateEndpointGroupResult

        val request = ResourceHandlerRequest.builder<ResourceModel>()
                .desiredResourceState(desiredModel)
                .previousResourceState(previousModel)
                .build()

        val response = UpdateHandler().handleRequest(proxy, request, null, logger)

        assertNotNull(response)
        assertEquals(OperationStatus.IN_PROGRESS, response.status)
        assertNotNull(response.resourceModel)
        assertNotNull(response.callbackContext)
        assertEquals(0, response.callbackDelaySeconds)
        assertEquals(endpointGroupArn, response.resourceModel.endpointGroupArn)
        assertTrue(response.callbackContext!!.pendingStabilization)
        assertNull(response.resourceModel.portOverrides)
        verify(exactly = 1) { proxy.injectCredentialsAndInvoke(ofType(), ofType<ProxyUpdateEndpointGroup>()) }
        assertNull(updateEndpointGroupRequestSlot.captured.portOverrides)
    }

    @Test
    fun handleRequest_UpdateEndpointGroupWithEmptyPortOverrides() {
        val portOverrides = ArrayList<PortOverride>()
        val previousModel = createTestResourceModel()

        val desiredModel = createTestResourceModel()
        desiredModel.portOverrides = portOverrides
        val describeEndpointGroupResult = DescribeEndpointGroupResult()
                .withEndpointGroup(EndpointGroup()
                        .withEndpointGroupArn(endpointGroupArn))
        every { proxy.injectCredentialsAndInvoke(ofType(), ofType<ProxyDescribeEndpointGroup>()) } returns describeEndpointGroupResult

        val updateEndpointGroupRequestSlot = slot<UpdateEndpointGroupRequest>()
	val updateEndpointGroupResult = UpdateEndpointGroupResult()
                .withEndpointGroup(EndpointGroup().withEndpointGroupArn(endpointGroupArn))
        every { proxy.injectCredentialsAndInvoke(capture(updateEndpointGroupRequestSlot), ofType<ProxyUpdateEndpointGroup>()) } returns updateEndpointGroupResult

        val request = ResourceHandlerRequest.builder<ResourceModel>()
                .desiredResourceState(desiredModel)
                .previousResourceState(previousModel)
                .build()

        val response = UpdateHandler().handleRequest(proxy, request, null, logger)

        assertNotNull(response)
        assertEquals(OperationStatus.IN_PROGRESS, response.status)
        assertNotNull(response.resourceModel)
        assertNotNull(response.callbackContext)
        assertEquals(0, response.callbackDelaySeconds)
        assertEquals(endpointGroupArn, response.resourceModel.endpointGroupArn)
        assertTrue(response.callbackContext!!.pendingStabilization)
        assertEquals(0, response.resourceModel.portOverrides.size)
        verify(exactly = 1) { proxy.injectCredentialsAndInvoke(ofType(), ofType<ProxyUpdateEndpointGroup>()) }
        assertEquals(ArrayList<String>(), updateEndpointGroupRequestSlot.captured.portOverrides) // Empty array to clean port-overrides.
    }

    @Test
    fun handleRequest_UpdateEndpointGroupWithUpdatedEndpointConfigurations() {
        val previousModel = createTestResourceModel()
        val desiredModel = createTestResourceModel()
        val endpointConfigurations = ArrayList<EndpointConfiguration>()
        endpointConfigurations.add(EndpointConfiguration.builder().endpointId("EPID1").weight(50).clientIPPreservationEnabled(false).build())
        endpointConfigurations.add(EndpointConfiguration.builder().endpointId("EPID2").weight(50).clientIPPreservationEnabled(false).build())
        endpointConfigurations.add(EndpointConfiguration.builder().endpointId("EPID3").weight(50).clientIPPreservationEnabled(true).build())
        desiredModel.endpointConfigurations = endpointConfigurations

        val describeEndpointGroupResult = DescribeEndpointGroupResult()
                .withEndpointGroup(EndpointGroup()
                        .withEndpointGroupArn(endpointGroupArn))
        every { proxy.injectCredentialsAndInvoke(ofType(), ofType<ProxyDescribeEndpointGroup>()) } returns describeEndpointGroupResult

        val updateEndpointGroupResult = UpdateEndpointGroupResult()
                .withEndpointGroup(EndpointGroup().withEndpointGroupArn(endpointGroupArn))
        every { proxy.injectCredentialsAndInvoke(ofType(), ofType<ProxyUpdateEndpointGroup>()) } returns updateEndpointGroupResult

        val request = ResourceHandlerRequest.builder<ResourceModel>()
                .desiredResourceState(desiredModel)
                .previousResourceState(previousModel)
                .build()

        val response = UpdateHandler().handleRequest(proxy, request, null, logger)

        assertNotNull(response)
        assertEquals(OperationStatus.IN_PROGRESS, response.status)
        assertNotNull(response.resourceModel)
        assertNotNull(response.callbackContext)
        assertEquals(0, response.callbackDelaySeconds)
        assertEquals(endpointGroupArn, response.resourceModel.endpointGroupArn)
        assertTrue(response.callbackContext!!.pendingStabilization)
        assertEquals(3, response.resourceModel.endpointConfigurations.size)
        assertEquals(response.resourceModel.endpointConfigurations, endpointConfigurations)
    }    
}
